File: ts/127.word-ladder.ts
/*
 * @lc app=leetcode id=127 lang=typescript
 *
 * [127] Word Ladder
 */

// @lc code=start
function ladderLength(
  beginWord: string,
  endWord: string,
  wordList: string[]
): number {
  const wordSet = new Set(wordList);

  let queue = [beginWord];
  let steps = 1;

  while (queue.length) {
    let nq = [];

    for (let word of queue) {
      if (word === endWord) return steps;

      for (let i = 0; i < word.length; i++) {
        for (let j = 0; j < 26; j++) {
          const nw =
            word.slice(0, i) + String.fromCharCode(97 + j) + word.slice(i + 1);

          if (wordSet.has(nw)) {
            nq.push(nw);
            wordSet.delete(nw);
          }
        }
      }
    }

    queue = nq;
    steps++;
  }

  return 0;
}
// @lc code=end

File: go/127.word-ladder/127.word-ladder.go
/*
 * @lc app=leetcode id=127 lang=golang
 *
 * [127] Word Ladder
 */

package leetcode

// @lc code=start
func ladderLength(beginWord string, endWord string, wordList []string) int {
	// word map to validate candidate
	wordMap := getWordMap(wordList, beginWord)

	// que for BFS
	que := []string{beginWord}
	depth := 0

	for len(que) > 0 {
		depth++

		n := len(que)
		for i := 0; i < n; i++ {
			word := que[0]
			que = que[1:]

			candidates := getCandidates(word)

			for _, c := range candidates {
				if _, ok := wordMap[c]; ok {
					if c == endWord {
						return depth + 1
					}

					delete(wordMap, c)
					que = append(que, c)
				}
			}
		}
	}

	return 0
}

func getWordMap(wordList []string, beginWord string) map[string]int {
	wordMap := make(map[string]int)

	for i, word := range wordList {
		if _, ok := wordMap[word]; !ok {
			if word != beginWord {
				wordMap[word] = i
			}
		}
	}

	return wordMap
}

func getCandidates(word string) []string {
	res := []string{}
	for i := 0; i < 26; i++ {
		for j := range word {
			new := replaceAt(word, j, rune('a'+i))
			if new != word {
				res = append(res, new)
			}
		}
	}
	return res
}
func replaceAt(target string, index int, new rune) string {
	runes := []rune(target)

	runes[index] = new

	return string(runes)
}

// @lc code=end
